sql/inspect: add job metrics for INSPECT jobs

spilchen opened the pull request:
Implements three metrics for INSPECT jobs:
- jobs.inspect.runs: Total number of INSPECT job executions
- jobs.inspect.runs_with_issues: Count of jobs that found inconsistencies
- jobs.inspect.issues_found: Aggregate count of issues found across all processors

Closes #153884

Epic: CRDB-30356
Release note: none

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154525)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @kev-cao)

___
*[`pkg/sql/inspect/inspect_metrics.go` line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154525#-OaW5f2PC10C9x8ssLAj:-OaW5f2PC10C9x8ssLAk:b413e6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/e489f55077970c333269ed84981714eaa6caa414/pkg/sql/inspect/inspect_metrics.go#L29)):*
> ```Go
> 
> var (
> 	metaInspectRuns = metric.Metadata{
> ```

the metrics look good, but there's some new additional guidance around adding these: https://cockroachlabs.atlassian.net/wiki/spaces/OI/pages/4592926723/Engineering+Playbook+CRDB+Metrics

i believe the only step we need from that page is to add the metrics to the datadog integration. (this is not a blocking comment, just a reminder)
___
*[`pkg/sql/inspect/inspect_processor.go` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154525#-OaW5LkIAwpbU9x1J8Gg:-OaW5LkIAwpbU9x1J8Gh:bohdbdq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e489f55077970c333269ed84981714eaa6caa414/pkg/sql/inspect/inspect_processor.go#L34)):*
> ```Go
> 	// errInspectFoundInconsistencies is a sentinel error used to mark errors
> 	// returned when INSPECT jobs find data inconsistencies.
> 	errInspectFoundInconsistencies = errors.New("INSPECT found inconsistencies")
> ```

could we just make this a `pgerror` and return this directly below? then we should be able to avoid needing to use `Mark` with a similar error object.
___
*[`pkg/sql/inspect/inspect_processor.go` line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154525#-OaW55pm1MBuCL7-cEwG:-OaW55pm1MBuCL7-cEwH:b-3fhmsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e489f55077970c333269ed84981714eaa6caa414/pkg/sql/inspect/inspect_processor.go#L206)):*
> ```Go
> 		},
> 		&metricsLogger{
> 			issuesFoundCtr: metrics.IssuesFound.AddChild(jobID.String()),
> ```

do other jobs have this pattern of adding a child metric per job ID? just curious, since it seems like the cardinality of that could get quite high, and for the most part i wouldn't expect people to need to filter by job ID here (since they are random IDs)


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154525)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @kev-cao, and @rafiss)

___
*[`pkg/sql/inspect/inspect_metrics.go` line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154525#-OaW5f2PC10C9x8ssLAj:-OaZfLAV41HFHWZhlHCb:bqkxf6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/e489f55077970c333269ed84981714eaa6caa414/pkg/sql/inspect/inspect_metrics.go#L29)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

the metrics look good, but there's some new additional guidance around adding these: https://cockroachlabs.atlassian.net/wiki/spaces/OI/pages/4592926723/Engineering+Playbook+CRDB+Metrics

i believe the only step we need from that page is to add the metrics to the datadog integration. (this is not a blocking comment, just a reminder)
</blockquote></details>

Thanks for letting me know. 
___
*[`pkg/sql/inspect/inspect_processor.go` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154525#-OaW5LkIAwpbU9x1J8Gg:-OaZfczW-U0rXXfOWIBx:b-g3tt63) ([raw file](https://github.com/cockroachdb/cockroach/blob/e489f55077970c333269ed84981714eaa6caa414/pkg/sql/inspect/inspect_processor.go#L34)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could we just make this a `pgerror` and return this directly below? then we should be able to avoid needing to use `Mark` with a similar error object.
</blockquote></details>

Yes, much cleaner.
___
*[`pkg/sql/inspect/inspect_processor.go` line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/154525#-OaW55pm1MBuCL7-cEwG:-OaZfeXN7TsV5o-N_pzE:b-na7pcg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e489f55077970c333269ed84981714eaa6caa414/pkg/sql/inspect/inspect_processor.go#L206)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

do other jobs have this pattern of adding a child metric per job ID? just curious, since it seems like the cardinality of that could get quite high, and for the most part i wouldn't expect people to need to filter by job ID here (since they are random IDs)
</blockquote></details>

No, there is no other pattern of using job ID as the child. I think I was misguided on this one. This metric is counted at each processor, and I originally went with an aggregated metric (`aggmetric.Counter`) thinking that this was how I can get a single value at the coordinator. But now that I understand this better, I believe that part isn't necessary. I have a test that checks this metric, but it's only 1-node (e.g. coordinator is one of the processors), so the aggregated part wasn't necessary for that either. I moved it back to a regular `metric.Counter`.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/inspect: add job metrics for INSPECT jobs

Implements three metrics for INSPECT jobs:
- jobs.inspect.runs: Total number of INSPECT job executions
- jobs.inspect.runs_with_issues: Count of jobs that found
  inconsistencies
- jobs.inspect.issues_found: Aggregate count of issues found across all
  processors

Closes #153884

Epic: CRDB-30356
Release note: none

rafiss reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/154525#-:-OaaLB9aBMJG918_XP2J:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi and @kev-cao)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+

spilchen commented:
bors r+
